1-source-code/1-workspace/2-tooling.md

```diff
@@ -0,0 +1,318 @@
+nc sync/atomic syscall time unicode/utf8 unsafe] false <nil> [] [] [] [] []}
+```
+
+By using the template language documented in the
+[`"text/template"`](https://golang.org/pkg/text/template) package, you can extract
+specific pieces of information.
+
+For instance, you could list all the import statements in `"github.com/golang/example/hello"`:
+
+```bash
+$ go list -f '{{.Imports}}' github.com/golang/example/hello
+[fmt github.com/golang/example/stringutil]
+```
+
+Once you learn a bit more of the Go templating language you'll be able to make your output look better:
+
+```bash
+$ go list -f '{{join .Imports "\n"}}' github.com/golang/example/hello
+fmt
+github.com/golang/example/stringutil
+```
+
+Alternatively, you can also get all of the information linked to package in JSON,
+by adding the `-json` flag.
+
+```bash
+$ go list -json github.com/golang/example/hello
+{
+	"Dir": "/Users/campoy/src/github.com/golang/example/hello",
+	"ImportPath": "github.com/golang/example/hello",
+	"Name": "main",
+	"Target": "/Users/campoy/bin/hello",
+	"Stale": true,
+	"StaleReason": "build ID mismatch",
+	"Root": "/Users/campoy",
+	"GoFiles": [
+		"hello.go"
+	],
+	"Imports": [
+		"fmt",
+		"github.com/golang/example/stringutil"
+	],
+	"Deps": [
+		"errors",
+		"fmt",
+		"github.com/golang/example/stringutil",
+		"internal/cpu",
+		"internal/poll",
+		"internal/race",
+		"io",
+		"math",
+		"os",
+		"reflect",
+		"runtime",
+		"runtime/internal/atomic",
+		"runtime/internal/sys",
+		"strconv",
+		"sync",
+		"sync/atomic",
+		"syscall",
+		"time",
+		"unicode/utf8",
+		"unsafe"
+	]
+}
+```
+
+Note that if you use the `-json` flag the `-f` flag will be ignored.
+
+### Exercise with `go list`
+
+Can you count how many `.go` files are under `github.com/golang/example` using `go list`?
+
+To see one possible solution, expand the `Details` below.
+
+<details>
+
+List all of the files with `go list`:
+
+```bash
+$ go list -f '{{join .GoFiles "\n"}}' github.com/golang/example/...
+app.go
+weave.go
+main.go
+main.go
+hello.go
+main.go
+main.go
+lookup.go
+main.go
+main.go
+main.go
+main.go
+hello.go
+main.go
+reverse.go
+main.go
+```
+
+If you're in Linux or Mac you can add `wc -l` to get a number:
+
+```bash
+$ go list -f '{{join .GoFiles "\n"}}' github.com/golang/example/... | wc -l
+16
+```
+</details>
+
+## Multiple paths in `GOPATH`
+
+In a similar way to how the `PATH` environment variable can contain multiple directories, which are used to find binaries
+in your system, `GOPATH` can contain multiple directories which will be used consecutively to find the packages you import.
+
+The `go` tool will try to find a package in the first component of `GOPATH`, and fallback to the second one only if the
+package wasn't found yet. Whenever you run `go get` the source code will be stored under the first `GOPATH` component.
+
+This setup is not used often, but it is important to know it exists. We will not cover it in any detail during the workshop.
```

Comment:
I'm gonna leave it in, for now. but might drop it depending on the feedback today